app/plugin_db_driver.go

```diff
@@ -0,0 +1,302 @@
+	if err != nil {
+		return "", err
+	}
+	connID := model.NewId()
+	d.connMut.Lock()
+	d.connMap[connID] = conn
+	d.connMut.Unlock()
+	return connID, nil
+}
+
+// According to https://golang.org/pkg/database/sql/#Conn, a client can call
+// Close on a connection, concurrently while running a query.
+//
+// Therefore, we have to handle the case where the connection is no longer
+// present in the map because it has been closed. ErrBadConn is a good choice
+// here which indicates the sql package to retry on a new connection.
+//
+// ConnPing, ConnQuery, ConnClose, Tx, and Stmt does this.
+
+func (d *DriverImpl) ConnPing(connID string) error {
+	d.connMut.RLock()
+	conn, ok := d.connMap[connID]
+	d.connMut.RUnlock()
+	if !ok {
+		return driver.ErrBadConn
+	}
+
+	return conn.Raw(func(innerConn interface{}) error {
+		return innerConn.(driver.Pinger).Ping(context.Background())
+	})
+}
+
+func (d *DriverImpl) ConnQuery(connID, q string, args []driver.NamedValue) (_ string, err error) {
+	var rows driver.Rows
+	d.connMut.RLock()
+	conn, ok := d.connMap[connID]
+	d.connMut.RUnlock()
+	if !ok {
+		return "", driver.ErrBadConn
+	}
+
+	err = conn.Raw(func(innerConn interface{}) error {
+		rows, err = innerConn.(driver.QueryerContext).QueryContext(context.Background(), q, args)
+		return err
+	})
+	if err != nil {
+		return "", err
+	}
+
+	rowsID := model.NewId()
+	d.rowsMut.Lock()
+	d.rowsMap[rowsID] = rows
+	d.rowsMut.Unlock()
+
+	return rowsID, nil
+}
+
+func (d *DriverImpl) ConnExec(connID, q string, args []driver.NamedValue) (_ plugin.ResultContainer, err error) {
+	var res driver.Result
+	var ret plugin.ResultContainer
+	d.connMut.RLock()
+	conn, ok := d.connMap[connID]
+	d.connMut.RUnlock()
+	if !ok {
+		return ret, driver.ErrBadConn
+	}
+
+	err = conn.Raw(func(innerConn interface{}) error {
+		res, err = innerConn.(driver.ExecerContext).ExecContext(context.Background(), q, args)
+		return err
+	})
+	if err != nil {
+		return ret, err
+	}
+
+	ret.LastID, ret.LastIDError = res.LastInsertId()
+	ret.RowsAffected, ret.RowsAffectedError = res.RowsAffected()
+
+	return ret, nil
+}
+
+func (d *DriverImpl) ConnClose(connID string) error {
+	d.connMut.Lock()
+	conn, ok := d.connMap[connID]
+	if !ok {
+		return driver.ErrBadConn
+	}
+	delete(d.connMap, connID)
+	d.connMut.Unlock()
+
+	return conn.Close()
+}
+
+func (d *DriverImpl) Tx(connID string, opts driver.TxOptions) (_ string, err error) {
+	var tx driver.Tx
+	d.connMut.RLock()
+	conn, ok := d.connMap[connID]
+	d.connMut.RUnlock()
+	if !ok {
+		return "", driver.ErrBadConn
+	}
+
+	err = conn.Raw(func(innerConn interface{}) error {
+		tx, err = innerConn.(driver.ConnBeginTx).BeginTx(context.Background(), opts)
+		return err
+	})
+	if err != nil {
+		return "", err
+	}
+
+	txID := model.NewId()
+	d.txMut.Lock()
+	d.txMap[txID] = tx
+	d.txMut.Unlock()
+	return txID, nil
+}
+
+func (d *DriverImpl) TxCommit(txID string) error {
+	d.txMut.Lock()
+	tx := d.txMap[txID]
```

Comment:
Are we accepting that a bad/invalid input could still cause a panic? Just making sure :)

Reply:
Transactions already have an atomic to track whether they are done or not in the sql package level. If a transaction is committed or rolled back after it's closed, the package returns `ErrTxDone`.

https://golang.org/src/database/sql/sql.go?s=57035:58170#L2044

Same for Rows and Stmt also.

Reply:
I see, I was actually worried about possible `nil` dereferences there.

Reply:
Resolved internally, I've also verified that's not the case so all good :+1: